packages/contractkit/src/identity/offchain/schema-utils.ts

```diff
@@ -10,34 +30,43 @@ export class SingleSchema<T> {
     readonly dataPath: string
   ) {}
 
-  async read(account: string) {
-    return readWithSchema(this.wrapper, this.type, account, this.dataPath)
+  async readAsResult(account: string) {
+    return readWithSchemaAsResult(this.wrapper, this.type, account, this.dataPath)
   }
 
+  read = makeAsyncThrowable(this.readAsResult.bind(this))
+
   async write(data: T) {
     return writeWithSchema(this.wrapper, this.type, this.dataPath, data)
   }
 }
 
-export const readWithSchema = async <T>(
+export const readWithSchemaAsResult = async <T>(
   wrapper: OffchainDataWrapper,
   type: t.Type<T>,
   account: Address,
   dataPath: string
-) => {
-  const data = await wrapper.readDataFrom(account, dataPath)
-  if (data === undefined) {
-    return
-  }
-  const asJson = JSON.parse(data)
-  const parseResult = type.decode(asJson)
-  if (isLeft(parseResult)) {
-    return undefined
+): Promise<Result<T, SchemaErrors>> => {
+  const resp = await wrapper.readDataFromAsResult(account, dataPath)
```

Comment:
something challenging about the overall solution is that `Promises` can fail... and that's a different error path to Result...

So if you're really defensive about your code, you should `try/catch` your awaits.... and have 2 error checking  mechanism.

Reply:
Promise errors vs. result errors is an interesting question but also not one I think we can generally give recommendations about

I.e. at the lowest levels, code calling IO or external code should indeed catch pretty much everywhere. How much a caller calling a result can rely on that is a more difficult of course. In this example, the readAsResult method certainly needs to catch lower exceptions